src/libs/globbing/globbing.c

```diff
@@ -0,0 +1,212 @@
+ * 	<li> if the pattern ends with "&#47;**", matching key names may contain arbitrary suffixes </li>
+ * </ul>
+ *
+ * @note '*' cannot match an empty key name part. This also means patterns like "something&#47;*" will
+ * not match the key "something". This is because each slash ('/') in the pattern has to correspond to
+ * a slash in the canonical key name, which neither end in a slash nor contain multiple slashes in sequence.
+ *
+ * @param key the Key to match against the globbing pattern
+ * @param pattern the globbing pattern used
+ * @retval 0 if @p key is not NULL, @p pattern is not NULL and @p pattern matches @p key
+ * @retval ELEKTRA_GLOB_NOMATCH otherwise
+ *
+ * @see isArrayName(), for info on valid array items
+ */
+int keyGlob (const Key * key, const char * pattern)
+{
+	if (key == NULL || pattern == NULL)
+	{
+		return ELEKTRA_GLOB_NOMATCH;
+	}
+
+	size_t nameSize = (size_t) keyGetNameSize (key);
+	char * name = elektraMalloc (nameSize);
+	keyGetName (key, name, nameSize);
+
+	const size_t patternSlashes = strcnt (pattern, '/');
+
+	size_t len = strlen (pattern);
+	bool prefixMode = len >= 2 && elektraStrCmp (pattern + len - 3, "/**") == 0;
+
+	char * patternEnd = name;
+	for (size_t i = 0; i < patternSlashes; ++i)
+	{
+		patternEnd = strchr (patternEnd + 1, '/');
+
+		if (patternEnd == NULL)
+		{
+			// more slashes in pattern, cannot match
+			free (name);
+			return ELEKTRA_GLOB_NOMATCH;
+		}
+	}
+
+	if (prefixMode)
+	{
+		// mark end of relevant part
+		*(patternEnd + 1) = '\0';
+	}
+	else if (strchr (patternEnd + 1, '/') != NULL)
+	{
+		// more slashes in name, cannot match
+		free (name);
+		return ELEKTRA_GLOB_NOMATCH;
+	}
+
+	char * fnmPattern = elektraToFnmatchGlob (elektraStrDup (pattern));
+	int rc = fnmatch (fnmPattern, name, FNM_PATHNAME);
+	elektraFree (fnmPattern);
+
+	if (rc == FNM_NOMATCH)
+	{
+		free (name);
+		return ELEKTRA_GLOB_NOMATCH;
+	}
+
+	const char * ptr = pattern;
+	const char * keyPtr = name;
+	while ((ptr = strchr (ptr + 1, '/')) != NULL && (keyPtr = strchr (keyPtr + 1, '/')) != NULL)
+	{
+		if (*(ptr + 2) == '/' || *(ptr + 2) == '\0')
+		{
+			if (*(ptr + 1) == '#' && !isArrayName (keyPtr + 1))
+			{
+				free (name);
+				return ELEKTRA_GLOB_NOMATCH;
+			}
+
+			if (*(ptr + 1) == '_' && isArrayName (keyPtr + 1))
+			{
+				free (name);
+				return ELEKTRA_GLOB_NOMATCH;
+			}
+		}
+	}
+
+	return 0;
+}
+
+/**
+ * @brief filters a given KeySet by applying a globbing pattern
+ *
+ * @param result the KeySet to which the matching keys should be appended
+ * @param input the KeySet whose keys should be filtered
+ * @param pattern the globbing pattern used
+ * @return the number of Keys appended to result or -1,
+ * 	   if @p result, @p input or @p pattern are NULL
+ *
+ * @see keyGlob(), for explanation of globbing pattern
+ */
+int ksGlob (KeySet * result, KeySet * input, const char * pattern)
```

Comment:
Return the result? (Like in ksCut?)

Reply:
This function is basically the same as `elektraKsFilter` with the filter function (in this case `keyGlob`) inlined.

We should either allocate a new KeySet with `ksNew` and return that or leave it as is, because in the current implementation you could pass a non-empty KeySet as `result` and get a number of appended Keys as the return value. Appending to a non-empty KeySet might be useful, if the result of several globbing operations are needed in a single KeySet, in which case this implementation would be more efficient than allocating multiple KeySets and merging them.

Reply:
Why not using elektraKsFilter? Then the dep to libease should be justified.

Did you measure the performance difference? It would be interesting.